ksql-engine/src/main/java/io/confluent/ksql/function/UdfCompiler.java

```diff
@@ -0,0 +1,138 @@
+      .put(Map.class, index -> typeConversionCode("Map", index))
+      .put(List.class, index -> typeConversionCode("List", index))
+      .build();
+
+
+  UdfInvoker compile(final Method method, final ClassLoader loader) {
+    try {
+      final IScriptEvaluator scriptEvaluator = createScriptEvaluator(method, loader);
+      final String code = generateCode(method);
+      return (UdfInvoker) scriptEvaluator.createFastEvaluator(code,
+          UdfInvoker.class, new String[]{"thiz", "args"});
+    } catch (final KsqlException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new KsqlException("Failed to compile udf wrapper class for "
+          + method, e);
+    }
+  }
+
+  private static String generateCode(final Method method) {
+    final StringBuilder builder = new StringBuilder();
+    final Class<?>[] params = method.getParameterTypes();
+    for (int i = 0; i < params.length; i++) {
+      final Function<Integer, String> converter = typeConverters.get(params[i]);
+      if (converter == null) {
+        throw new KsqlException("Type " + params[i] + " is not supported in UDFs");
+      }
+      builder.append(converter.apply(i)).append("\n");
+    }
+
+    builder.append("\nreturn ((").append(method.getDeclaringClass().getSimpleName())
+        .append(") thiz).").append(method.getName()).append("(");
+
+    for (int i = 0; i < params.length; i++) {
+      builder.append("arg").append(i).append(",");
+    }
+
+    builder.deleteCharAt(builder.length() - 1);
+    builder.append(");");
+    final String code = builder.toString();
+    logger.debug("generated code for udf method = {}\n{}", method, code);
+    return code;
+  }
+
+  private static IScriptEvaluator createScriptEvaluator(final Method method,
+                                                        final ClassLoader loader) throws Exception {
+    final IScriptEvaluator scriptEvaluator
+        = CompilerFactoryFactory.getDefaultCompilerFactory().newScriptEvaluator();
+    scriptEvaluator.setClassName(method.getDeclaringClass().getName() + "_" + method.getName());
+    final String udfClassName = method.getDeclaringClass().getName();
+    scriptEvaluator.setDefaultImports(new String[]{
+        "java.util.*",
+        "io.confluent.ksql.function.KsqlFunctionException",
+        udfClassName,
+    });
+    scriptEvaluator.setParentClassLoader(loader);
+    return scriptEvaluator;
+  }
+
+  private static String typeConversionCode(final String type, final int index) {
+    if (type.equals("Map") || type.equals("List")) {
+      return type + " arg" + index + " = (" + type + ")args[" + index + "];\n";
+    }
+    final String argArrayVal = "args[" + index + "]";
+    final String argVarAssignment = "arg" + index + " = ";
+    final StringBuilder builder = new StringBuilder();
+    final String numericValue = type.equals("Integer") ? "intValue()" : type.toLowerCase()
+        + "Value()";
+    builder.append(type).append(" arg").append(index).append(";\n")
```

Comment:
May help to put a comment with the actually generated code without the string builder boiler plate. The latter makes this very hard to read.

Reply:
you mean a comment in this class that shows what is generated?

Reply:
Yea. Or just for this method. The generated code follows a very simple template, but that is hard to parse due to the builder usages.

Reply:
Ah. You added some code on line 74. that is sufficient IMO, as it covers the code generated by this method.